### DISCUSSION TITLE
Validating Text inputs with currency formatting

### DISCUSSION THREAD

**Discussion by abpbackup** [colinhacks/zod#3262] [Q&A] [answered]
@JacobWeisenburger @rubengmurray , I have this case where I need the type to be a number for my Type in typescript. However, since I'm using a TextInput which only accepts strings, and I'm also formatting the number to currency using NumberFormat lib from react-number-format, I need to remove all the format before so that I can trigger a min/max validation. This code snipped properly removes all the format and leaves the plain number but the min/max validation is ignored. Any ideas?

value: z.preprocess((x) => {
const val = x ? String(x) : '0';
const numberVal = Number(val.replace(/[^0-9\.-]+/g, ''));
return numberVal;
}, z.number().min(0.0001).max(999999999)),

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const schema = z.union( [
    z.string().transform( x => x.replace( /[^0-9.-]+/g, '' ) ),
    z.number(),
] ).pipe( z.coerce.number().min( 0.0001 ).max( 999999999 ) )

type Input = z.input<typeof schema>
// type Input = string | number

type Output = z.output<typeof schema>
// type Output = number

console.log( schema.safeParse( 0.0001 ).success ) // true
console.log( schema.safeParse( 0.00001 ).success ) // false

console.log( schema.safeParse( 999999999 ).success ) // true
console.log( schema.safeParse( 9999999999 ).success ) // false

console.log( schema.parse( 42.42 ) ) // 42.42
console.log( schema.parse( '42.42' ) ) // 42.42
console.log( schema.parse( '$42.42' ) ) // 42.42
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
